xcode/Subconscious/Shared/UIKit/BlockEditor/BlockEditorViewController.swift

```diff
@@ -0,0 +1,1211 @@
+                state: state,
+                id: id,
+                selection: selection
+            )
+        }
+    }
+    
+    func textDidChange(
+        state: Model,
+        id: UUID?,
+        text: String,
+        selection: NSRange
+    ) -> Update {
+        guard let id = id else {
+            Self.logger.log("No block id. Doing nothing.")
+            return Update(state: state)
+        }
+        guard let i = state.blocks.firstIndex(whereID: id) else {
+            Self.logger.log("block#\(id) not found. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        var model = state
+        let block = state.blocks[i].setText(text: text, selection: selection)
+        
+        guard let block = block else {
+            Self.logger.log("block#\(id) could not update block text. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        model.blocks[i] = block
+        
+        return Update(state: model)
+    }
+    
+    func didChangeSelection(
+        state: Model,
+        id: UUID,
+        selection: NSRange
+    ) -> Update {
+        guard let i = state.blocks.firstIndex(whereID: id) else {
+            Self.logger.log("block#\(id) not found. Doing nothing.")
+            return Update(state: state)
+        }
+        var model = state
+        guard let block = state.blocks[i].setSelection(
+            selection: selection
+        ) else {
+            Self.logger.log("block#\(id) could not update block selection. Doing nothing.")
+            return Update(state: state)
+        }
+        model.blocks[i] = block
+        return Update(state: model)
+    }
+    
+    func splitBlock(
+        state: Model,
+        id: UUID,
+        selection nsRange: NSRange
+    ) -> Update {
+        guard let indexA = state.blocks.firstIndex(whereID: id) else {
+            Self.logger.log("block#\(id) not found. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        Self.logger.log("block#\(id) splitting at \(nsRange.location)")
+        
+        let blockA = state.blocks[indexA]
+        
+        guard let blockTextA = blockA.text else {
+            Self.logger.log("block#\(id) cannot split block without text. Doing nothing.")
+            return Update(state: state)
+        }
+        
+        guard let (textA, textB) = blockTextA.splitAtRange(nsRange) else {
+            Self.logger.log(
+                "block#\(id) could not split text at range. Doing nothing."
+            )
+            return Update(state: state)
+        }
+        
+        guard let blockA = blockA.setText(
+            text: textA,
+            selection: NSRange(location: nsRange.location, length: 0)
+        ) else {
+            Self.logger.log(
+                "block#\(id) could set text. Doing nothing."
+            )
+            return Update(state: state)
+        }
+
+        var blockB = BlockEditor.TextBlockModel()
```

Comment:
Shouldn't we use whatever model `blockA` was rather than using `TextBlockModel` every time?